source/client_channel_handler.c

```diff
@@ -111,25 +122,52 @@ static int s_packet_handler_connack(
         /* If successfully connected, schedule all pending tasks */
         AWS_LOGF_TRACE(
             AWS_LS_MQTT_CLIENT, "id=%p: connection was accepted processing offline requests.", (void *)connection);
-        /**
-         * TODO: if the clean_session is true, the pending requests should be cleaned up.
-         */
+
         if (!aws_linked_list_empty(&requests)) {
+
             struct aws_linked_list_node *current = aws_linked_list_front(&requests);
             const struct aws_linked_list_node *end = aws_linked_list_end(&requests);
 
             do {
-
                 struct aws_mqtt_request *request = AWS_CONTAINER_OF(current, struct aws_mqtt_request, list_node);
-                AWS_LOGF_TRACE(
-                    AWS_LS_MQTT_CLIENT,
-                    "id=%p: processing offline request %" PRIu16,
-                    (void *)connection,
-                    request->packet_id);
-                aws_channel_schedule_task_now(connection->slot->channel, &request->outgoing_task);
-
+                if (connection->clean_session) {
+                    /* discard the previous session, invoke the complete callbacks first */
```

Comment:
We probably want to cancel existing requests when the user calls connect(), instead of when the connack comes back.

That makes it a little more clear that old requests are being cancelled because of the USER'S request for a clean session.

Also, it's a little confusing to let someone call connect(), publish(A), publish(B), and then A and B fail when the connection finally succeeds